docs/proposed/psa-conditional-inclusion-c.md

```diff
@@ -0,0 +1,198 @@
+
+### Requirements
+
+[Req.interface] The application can declare which cryptographic mechanisms it needs.
+
+[Req.inclusion] If the application does not require a mechanism, a suitably configured Mbed TLS build must not include it. The granularity of mechanisms must work for typical use cases and has [acceptable limitations](#acceptable-limitations).
+
+[Req.drivers] If a PSA driver is available in the build, a suitably configured Mbed TLS build must not include the corresponding software code (unless a software fallback is needed).
+
+[Req.c] The configuration mechanism consists of C preprocessor definitions, and the build does not require tools other than a C compiler. This is necessary to allow building an application and Mbed TLS in development environments that do not allow third-party tools.
+
+[Req.adaptability] The implementation of the mechanism must be adaptable with future evolution of the PSA cryptography specifications and Mbed TLS. Therefore the interface must remain sufficiently simple and abstract.
+
+### Acceptable limitations
+
+[Limitation.matrix] If a mechanism is defined by a combination of algorithms and key types, for example a block cipher mode (CBC, CTR, CFB, …) and a block permutation (AES, CAMELLIA, ARIA, …), there is no requirement to include only specific combinations.
+
+[Limitation.direction] For mechanisms that have multiple directions (for example encrypt/decrypt, sign/verify), there is no requirement to include only one direction.
+
+[Limitation.size] There is no requirement to include only support for certain key sizes.
+
+[Limitation.multipart] Where there are multiple ways to perform an operation, for example single-part and multi-part, there is no mechanism to select only one or a subset of the possible ways.
+
+## Interface
+
+### PSA Crypto configuration file
+
+The PSA crypto configuration file `psa/crypto_config.h` defines a series of symbols of the form `PSA_WANT_xxx` where `xxx` . The symbols are documented in the section [“PSA Crypto configuration symbols”](#psa-crypto-configuration-symbols) below.
+
+The symbol `MBEDTLS_PSA_CRYPTO_CONFIG` in `mbedtls/config.h` determines whether `psa/crypto_config.h`. is used.
+
+* If `MBEDTLS_PSA_CRYPTO_CONFIG` is unset, which is the default at least in Mbed TLS 2.x versions, things are as they are today: the PSA subsystem includes generic code unconditionally, and includes support for specific mechanisms conditionally based on the existing `MBEDTLS_xxx_` symbols.
+* If `MBEDTLS_PSA_CRYPTO_CONFIG` is set, the necessary software implementations of cryptographic algorithms are included based on both the content of the PSA crypto configuration file and the Mbed TLS configuration file. For example, the code in `aes.c` is enabled if either `mbedtls/config.h` contains `MBEDTLS_AES_C` or `psa/crypto_config.h` contains `PSA_WANT_KEY_TYPE_AES`.
```

Comment:
In practice, this means 'wants' from PSA Crypto configuration will only be able to add (include more functionality) to the mbedTLS config, right? So the Crypto config would be included as part of the mbedTLS config header, after the user config, but before check-config?

Reply:
Yes.

If an application uses a library that calls `mbedtls_aes_xxx` and another library that calls `psa_xxx(PSA_KEY_TYPE_AES)`, it should pull in `MBEDTLS_AES_C` for the sake of the first library and `PSA_KEY_TYPE_AES` for the sake of the second library. The fact that `PSA_WANT_KEY_TYPE_AES` currently implies `MBEDTLS_AES_C` is an implementation detail that may change if later `mbedtls_aes_xxx` becomes a wrapper around a PSA implementation.

I can only see two viable alternatives here: either the PSA configuration can add to the ”legacy“ configuration, or it has to be compatible (so if you have `PSA_WANT_KEY_TYPE_AES` but no PSA AES accelerator and no `MBEDTLS_AES_C`, you get a compile-time error). I proposed the add model because it's a lot more transparent for users: you get what you ask for.

The compatible model has the advantage that it's likely easier to implement, and perhaps easier to understand for the minority of users who care about both crypto APIs, because it's the way Mbed TLS works now: if you say you want X and you don't say you want Y, but X requires Y, you get a compile-time error from `check_config.h`. We'd like to move away from this model in Mbed TLS (hopefully it'll be one of the 3.0 changes), and instead go for the model where you get what you ask for.

I don't think the add model will be substantially harder to implement because it's about as much effort to write the dependencies in add form (if defined(X) define Y) as in compatible form (if defined(X) and not defined(Y) error).

Reply:
I think another advantage of the "add" model is that in case someone just wants to use the PSA Crypto API and doesn't care about any other part of Mbed TLS, they could use an empty `mbedtls/config.h` and only populate `psa/crypto_config.h` to get only what they want from PSA Crypto, without needing to know anything about Mbed TLS configuration. (With a "compatible" model, those people would need to know which Mbed TLS symbols correspond to the PSA symbols, even though they only care about PSA.)